File: .mise/lib/imports/diff.py
"""
The diff between a module's two versions, meaning the text read once per
path, the line matcher built once per pair, the original rows a formatted
row came from, and the hunk around a row.
"""

from difflib   import SequenceMatcher
from functools import cache
from pathlib   import Path

CONTEXT = 3


def hunk(pairs: SequenceMatcher, row: int | None, name: str = "") -> list[str]:
    """
    Return the unified-diff lines of `pairs` cut to `CONTEXT` lines either
    side of row `row` of its second sequence, or where the row is unknown
    of the first changed line naming `name`, else the first changed line, an
    ellipsis marking each cut.
    """
    shown = []
    for tag, i1, i2, j1, j2 in pairs.get_opcodes():
        if tag in ("delete", "replace"):
            shown += [(None, f"-{line}") for line in pairs.a[i1:i2]]
        if tag != "delete":
            mark   = " " if tag == "equal" else "+"
            shown += [(j1 + k, f"{mark}{line}") for k, line in enumerate(
                pairs.b[j1:j2]
            )]
    if row is None:
        changed = [k for k, (_, line) in enumerate(shown) if line[0] != " "]
        index   = next(iter([k for k in changed if name in shown[k][1]] or changed), 0)
    else:
        index = next((k for k, (seen, _) in enumerate(shown) if seen == row - 1), 0)
    low, high = max(0, index - CONTEXT), index + CONTEXT + 1
    return [
        *(["..."] if low else []),
        *(line for _, line in shown[low:high]),
        *(["..."] if high < len(shown) else [])
    ]


def mapped_rows(pairs: SequenceMatcher, row: int) -> range:
    """
    Return the rows of `pairs`'s first sequence that row `row` of its second
    came from, one row for an equal block, every row of the block that
    rewrote it, and the row an insertion landed at.
    """
    for tag, i1, i2, j1, j2 in pairs.get_opcodes():
        if j1 <= row - 1 < j2:
            if tag == "equal":
                return range(landed := i1 + row - j1, landed + 1)
            return range(i1 + 1, max(i1 + 1, i2) + 1)
    return range(0)


@cache
def pairing(after: Path, before: Path) -> SequenceMatcher:
    """
    Return the line matcher between the modules at `before` and `after`,
    built once per pair.
    """
    return SequenceMatcher(
        None,
        text_of(before).splitlines(),
        text_of(after).splitlines(),
        autojunk = False
    )


@cache
def text_of(path: Path) -> str:
    """
    Return the text of the module at `path`, read once per path.
    """
    return path.read_text(encoding="utf-8", errors="replace")

File: crate/tests/imports/diff.rs
//! The diff between a module's two versions, meaning the original rows a
//! formatted row came from and the hunk around a row.

use std::ops::Range;

use itertools::Itertools;
use similar::{ChangeTag, DiffTag, TextDiff};

/// How many diff lines a hunk shows either side of the row it centres on.
const CONTEXT: usize = 3;

/// The unified-diff lines around one row, cut to [`CONTEXT`] lines either
/// side with an ellipsis marking each cut.
///
/// Where the row is unknown the window centres on the first changed line
/// naming `name`, and failing that on the first changed line at all.
pub(crate) fn hunk(diff: &TextDiff<'_, '_, str>, row: Option<usize>, name: &str) -> Vec<String> {
    let shown: Vec<_> = diff
        .iter_all_changes()
        .map(|change| {
            let mark = match change.tag() {
                ChangeTag::Delete => '-',
                ChangeTag::Equal => ' ',
                ChangeTag::Insert => '+',
            };
            (change.new_index(), mark, change.value())
        })
        .collect();
    let index = row.map_or_else(
        || {
            let changed = || shown.iter().positions(|(_, mark, _)| *mark != ' ');
            changed()
                .find(|&at| !name.is_empty() && shown[at].2.contains(name))
                .or_else(|| changed().next())
                .unwrap_or_default()
        },
        |row| {
            shown
                .iter()
                .position(|(seen, _, _)| *seen == Some(row - 1))
                .unwrap_or_default()
        },
    );
    let low = index.saturating_sub(CONTEXT);
    let high = (index + CONTEXT + 1).min(shown.len());
    let opening = (low > 0).then(|| "...".to_owned());
    let closing = (high < shown.len()).then(|| "...".to_owned());
    opening
        .into_iter()
        .chain(
            shown[low..high]
                .iter()
                .map(|(_, mark, line)| format!("{mark}{line}")),
        )
        .chain(closing)
        .collect()
}

/// The original rows that one row of the formatted side came from, which is
/// one row for an unchanged block, every row of the block that rewrote it,
/// and the row an insertion landed at.
pub(crate) fn mapped_rows(diff: &TextDiff<'_, '_, str>, row: usize) -> Range<usize> {
    for op in diff.ops() {
        let (tag, old, new) = op.as_tag_tuple();
        if new.contains(&(row - 1)) {
            if tag == DiffTag::Equal {
                let landed = old.start + row - new.start;
                return landed..landed + 1;
            }
            return old.start + 1..(old.start + 1).max(old.end) + 1;
        }
    }
    0..0
}
